src/Core/Common/ImageBufferAndIndexRange/Code.cxx

```diff
@@ -0,0 +1,78 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+
+#include "itkImage.h"
+#include "itkImageBufferRange.h"
+#include "itkIndexRange.h"
+#include "itkNumericTraits.h"
+
+#include <numeric> // For iota
+
+namespace
+{
+// Creates an image with sequentially increasing pixel values (0, 1, 2, ...).
+template <typename TImage>
+typename TImage::Pointer
+CreateImageWithSequentiallyIncreasingPixelValues(const typename TImage::RegionType & region)
+{
+  using PixelType = typename TImage::PixelType;
+
+  const auto image = TImage::New();
+  image->SetRegions(region);
+  image->Allocate();
+
+  const itk::ImageBufferRange<TImage> imageBufferRange(*image);
+  std::iota(imageBufferRange.begin(), imageBufferRange.end(), PixelType{ 0 });
+
+  return image;
+}
+
+
+// Prints all pixel values with their N-dimensional indices.
+template <typename TImage>
+void
+PrintPixelValues(const TImage & image)
+{
+  const typename TImage::RegionType                        region = image.GetBufferedRegion();
+  const itk::ImageRegionIndexRange<TImage::ImageDimension> indexRange(region);
```

Comment:
I am less familiar with `itk::ImageRegionIndexRange`, could you please comment on the advantage of this class over [`itk::ImageRegionIteratorWithIndex`](https://itk.org/Doxygen/html/classitk_1_1ImageRegionIteratorWithIndex.html)?

Reply:
That's a very good question @tbirdso. Indeed, there is some overlap between the functionality of ImageRegionIndexRange and ImageRegionIteratorWithIndex. The most significant advantage of ImageRegionIndexRange is that it supports modern C++ iteration, including "range-based for loops". As included with the example:

    for (const auto & index : indexRange)
    {
      std::cout << "Index: " << index << '\n';
    }

Prints all the (N-dimensional) indices of pixels within the specified ImageRegion.

Another significant difference is that ImageRegionIndexRange is a more "lightweight" object than ImageRegionIteratorWithIndex. (Especially because `ImageRegionIndexRange` does not have a virtual function table, and it does not have a pointer to the image) So it might yield a better runtime performance. Especially for pixel-wise operations, this might be of interest.

Does that answer your question?

Reply:
Yes it does, very neat! Thank you 😄